processing/src/main/java/org/apache/druid/segment/column/Types.java

```diff
@@ -112,4 +113,12 @@ public static <T extends TypeDescriptor> boolean either(
     return (typeSignature1 != null && typeSignature1.is(typeDescriptor)) ||
            (typeSignature2 != null && typeSignature2.is(typeDescriptor));
   }
+
+  public static class IncompatibleTypeException extends IAE
```

Comment:
Q - would this exception return a 5xx to the user? I am wondering if we change the exception so that the end user gets a 400 instead. since its really a bad query.

Reply:
hmm, in this particular PR the thing isn't user facing really because its happening inside the brokers segment metadata cache that powers the SQL schema and is happening in the background. This logic does not happen in user queries.

The callers of `ExpressionTypeConversion` that also now throw this exception instead of `IAE` are i guess potentially user facing (though could also happen in ingest time transforms), though im not really sure this is an exception that should be thrown directly to users, rather it should probably be caught by something and decorated with additional context to indicate like what the offending expression was, etc.